File: diagnostics-web/src/app/Model/DiagProcess.ts
import 'reflect-metadata';
import {Null} from '../util/Null';
import * as _ from 'lodash';
import {Type} from 'class-transformer';
import {OnlineState} from './OnlineState';
import {Level} from './Level';

export class DiagProcess {
  id: string = '';
  processName: string = '';
  processId: number = 0;
  instanceName: Null<string> = '';
  machineName: string = '';
  userName: string = '';
  uri: string = '';
  message: string = '';
  alertLevel = 0;
  labelClass = '';

  @Type(() => Date)
  lastOnline: Null<Date> = null;
  state: OnlineState = 'NA';
  connectionId: string = '';


  constructor(data?: DiagProcess) {
    if (data)
      this.update(data);
  }

  get title(): string
  {
    return `${this.machineName}/${this.userName}/${this.processName}`;
  }

  update(s: DiagProcess): void {
    _.assign(this, s);
    //console.log(`${this.id} state: ${this.state}`)
    this.labelClass = this.alertLevel === 0 ? '' : 'event-level-' + Level.LevelToString(this.alertLevel).toLocaleLowerCase();
  }
}

File: Diagnostics.Service.Common/Common/DiagProcess.cs
#region Copyright

// Diagnostic Explorer, a .Net diagnostic toolset
// Copyright (C) 2010 Cameron Elliot
// 
// This file is part of Diagnostic Explorer.
// 
// Diagnostic Explorer is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Diagnostic Explorer is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public License
// along with Diagnostic Explorer.  If not, see <http://www.gnu.org/licenses/>.
// 
// http://diagexplorer.sourceforge.net/

#endregion

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text.Json.Serialization;
using System.Xml.Linq;
using System.Text.RegularExpressions;
using System.Xml.Serialization;
using DiagnosticExplorer.Util;
using MongoDB.Bson.Serialization.Conventions;

namespace DiagnosticExplorer.Common;

public class DiagProcess
{
    public string Id { get; set; }

    public string? InstanceId { get; set; }

    public string? ProcessName { get; set; }

    public int ProcessId { get; set; }

    public string? MachineName { get; set; }

    public string? UserName { get; set; }
        
    public string? Uri { get; set; }

    public OnlineState State { get; set; }

    public RegistrationMode RegistrationMode { get; set; } = RegistrationMode.NA;

    public string? Message { get; set; }
    public int AlertLevel { get; set; } = 0;


    [JsonIgnore] public DateTime? AlertLevelDate { get; set; } = null;

    public DateTime? LastOnline { get; set; }

    [Newtonsoft.Json.JsonIgnore]
    [JsonIgnore]
    public string? ConnectionId { get; set; }

 
    public DiagProcess Clone()
    {
        return (DiagProcess)MemberwiseClone();
    }

    public override string ToString()
    {
        return
            $"DiagProcess PID:{ProcessId} Process:{ProcessName}, Uri:{Uri}, State:{State}, LastOnline:{LastOnline}";
    }
}

public static class DiagProcessExtensions
{
    public static DiagProcess? FindById(this IEnumerable<DiagProcess> list, string id)
    {
        return list.FirstOrDefault(item => string.Equals(item.Id, id, StringComparison.InvariantCultureIgnoreCase));
    }

    public static DiagProcess? FindByConnectionId(this IEnumerable<DiagProcess> list, string connectionId)
    {
        return list.FirstOrDefault(item => string.Equals(item.ConnectionId, connectionId, StringComparison.InvariantCultureIgnoreCase));
    }

    public static DiagProcess? FindByInstanceId(this IEnumerable<DiagProcess> list, string instanceId)
    {
        return list.FirstOrDefault(item => string.Equals(item.InstanceId, instanceId, StringComparison.InvariantCultureIgnoreCase));
    }

    public static DiagProcess? FindByUri(this IEnumerable<DiagProcess> list, string uri)
    {
        return list.FirstOrDefault(item => string.Equals(item.Uri, uri, StringComparison.InvariantCultureIgnoreCase));
    }
}